graylog2-web-interface/src/components/common/SourceCodeEditor.jsx

```diff
@@ -0,0 +1,205 @@
+import React from 'react';
+import { PropTypes } from 'prop-types';
+import { Resizable } from 'react-resizable';
+import 'brace';
+import AceEditor from 'react-ace';
+import { Button, ButtonGroup, ButtonToolbar, OverlayTrigger, Tooltip } from 'react-bootstrap';
+
+import { ClipboardButton } from 'components/common';
+
+import 'brace/mode/text';
+import 'brace/theme/tomorrow';
+import 'brace/theme/monokai';
+import style from './SourceCodeEditor.css';
+
+/**
+ * Component that renders a source code editor input. This is what powers the pipeline rules and collector
+ * editors.
+ */
+class SourceCodeEditor extends React.Component {
+  static propTypes = {
+    /**
+     * Annotations to show in the editor's gutter. The format should be:
+     * `[{ row: 0, column: 2, type: 'error', text: 'Some error.'}]`
+     * The type value must be one of `error`, `warning`, or `info`.
+     */
+    annotations: PropTypes.array,
```

Comment:
Is it worth using a `PropTypes.shape` here or do we not bother because this typically comes from the backend anyway?

Reply:
I'm most of the time against those shape objects, as you need to define the whole schema of an object (or an array of objects in this case) in a way that is not very readable and it's easy to make you waste more time removing warnings than helping catching errors. Additionally, in this case the shape is highly dependant on ace, and they don't even have good documentation for it. For instance, the `column` property didn't seem to have any effect when I tested it, as all annotations appear in the gutter on the right of the editor. I also had to go to the source code to see what types were supported.
In my opinion, having an example and the vales the `type` property accepts is good enough.

Reply:
Works for me :)